full-stack-practice - Daniel Yount

TheDanYount opened the pull request:
![Recording 2024-09-17 at 15 11 38](https://github.com/user-attachments/assets/14b3218b-8fd3-48be-a462-e758f5efdcd4)

TheDanYount committed:
Practiced full stack fetching

RobertGardner commented on full-stack-practice/client/src/Catalog.tsx:
```diff
@@ -0,0 +1,66 @@
+      } catch (error) {
+        console.error(error);
+        alert('Error. Click "OK" to reload catalog');
+        setProducts(undefined); // this will force a re-render
```
I like the comment.

Since this only runs once on first mount, I believe products is already undefined, so I'm not sure this will cause a re-render. The best way to force the re-render is with an `isLoading` flag.

RobertGardner commented on full-stack-practice/client/src/Catalog.tsx:
```diff
@@ -0,0 +1,66 @@
+import { toDollars } from './to-dollars';
+
+export function Catalog() {
+  const [products, setProducts] = useState<Product[]>();
```
You assume that if `products` is not defined then it is loading. That isn't true if there is a network error. In that case, you will have already determined there's a problem, but `products` will be undefined, so you'll display a Loading message. That can be confusing to the user.

This is a nice thought, but I think the `isLoading` state is the best way to build this.

RobertGardner commented on full-stack-practice/client/src/Details.tsx:
```diff
@@ -0,0 +1,70 @@
+          </button>
+        </>
+      ) : (
+        <p className="p-4">Loading...</p>
```
This has the same issues as outlined in Catalog.

RobertGardner commented on full-stack-practice/client/src/Details.tsx:
```diff
@@ -0,0 +1,70 @@
+      } catch (error) {
+        console.error(error);
+        alert('Error. Click "OK" to return to catalog');
+        navigate('/');
```
This is reasonable.

RobertGardner reviewed (approved):
Well done. Some comments that might help improve the code.

TheDanYount committed:
Bug fix

I no longer conflate having a product/products loaded with loading. i.e. if there's an error that's handled separately.
